fixed gitlab integration

Zveroboi123 opened the pull request:
1. gitlab groups were fixed 
2. gitlab authentication fixed - gitlab expects only auth token. no username required

dhui commented on source/gitlab/gitlab.go:
```diff
@@ -8,11 +8,10 @@ import (
 	"strings"
-)
 
-import (
```
Leave the import grouping as is

Zveroboi123 replied:
Sure

Zveroboi123 commented:
Empty or dummy username would work just fine, of course, but might confuse users. The main fix was around the GitLab groups. Would you like to leave the previous format - "username@password" in place?

dhui commented:
> Would you like to leave the previous format - "username:password" in place?

Yes, leave the original URI format in place.

> The main fix was around the GitLab groups.

Your change is backwards incompatible. Have you tried URL encoding the `/` for your group?

Zveroboi123 commented:
I tried encoding it, but it gets automatically decoded and treated as regular slash. using backslash didn't work either.
I'd like to propose a new backward compatible format: gitlab://username:token@<domain>/<group>\\<project>/<sub-folder>
so that it would work either way. We could think of another group-project delimiter, if you like. 
What do you think?

Zveroboi123 commented:
Should I take care of the coverage degradation?
It's just 0.08% and is probably due to adding a single line of code.

dhui commented:
> I tried encoding it, but it gets automatically decoded and treated as regular slash. using backslash didn't work either.
I'd like to propose a new backward compatible format: gitlab://username:token@<domain>/<group>\\<project>/<sub-folder>
so that it would work either way. We could think of another group-project delimiter, if you like.
What do you think?

What's the valid character set for a username or group? I couldn't find any documentation about it.
This was the closest I could find: https://docs.gitlab.com/ee/user/reserved_names.html
If we're going to pick a character in the valid character set, we'll need to be able to escape it...
e.g. `\/` => `/` and `\\` => `\`

> Should I take care of the coverage degradation?

Thanks for being aware of this. I wouldn't worry about the decrease in code coverage at this point.

Zveroboi123 commented:
Got your point. I added escaping support and verified it is backward compatible.
Following formats will work fine:
1. gitlab://aaa:kZX7yRFHa8AMyKjYtTc3@gitlab.contoso.com/my\\\\group\\myproject/myprojectpath#master
2. gitlab://aaa:kZX7yRFHa8AMyKjYtTc3@gitlab.contoso.com/mygroup\\myproject/myprojectpath#master
3. 
gitlab://aaa:kZX7yRFHa8AMyKjYtTc3@gitlab.contoso.com/myproject/myprojectpath#master

dhui commented:
Actually, instead adding our own escaping to the DSN, use [url.URL.RawPath](http://localhost:8088/pkg/net/url/#URL) instead of `url.URL.Path`. That way, URL-encoded strings in the path should work.

Zveroboi123 commented:
Hi, I don't fully understand you. I'm not inventing my own escaping here.
I'm using regex to split group + projectId from the rest of the path. All
unescaping is done automatically. Just run my code to see how it works.

On Wed, Mar 13, 2019, 1:13 AM Dale Hui <notifications@github.com> wrote:

> Actually, instead adding our own escaping to the DSN, use url.URL.RawPath
> <http://localhost:8088/pkg/net/url/#URL> instead of url.URL.Path. That
> way, URL-encoded strings in the path should work.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/golang-migrate/migrate/pull/188#issuecomment-472216528>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Are_Lx0Z8EPPjd1TrnT6wv1YjcHbDQZmks5vWDSmgaJpZM4bphdL>
> .
>

dhui commented:
> Hi, I don't fully understand you. I'm not inventing my own escaping here.

The current approach of using regexs is not very clean. My original though was that we'd be adding backslash escaping on top of url encoding which is also dirty...
I believe both of these dirty approaches can be avoided by using the escaped path.

> Please check out the following thread: https://stackoverflow.com/questions/39270633/why-does-url-parse-not-populate-url-rawpath#

Ahh, yeah, it looks like we should be using `EscapedPath()` instead.
https://play.golang.org/p/V838Xn_OZMa

so `pe := strings.Split(strings.Trim(u.Path, "/"), "/")` would become `pe := strings.Split(strings.Trim(u.EscapedPath(), "/"), "/")`

> current gitlab implementation is pretty much useless for 95% of us.

No idea about the data for this... e.g. number of users using groups/projects Support for gitlab sources was only merged ~2 days ago.
Obviously, the original author didn't have this problem.

> it would be super cool if you would approve it as it already is to unblock us all.

If you're blocked, use your fork.

Also, **whatever the final solution is, we should document it** so future users aren't confused.

Zveroboi123 commented:
Hi,
I personally don't like the approach to have a slash delimiter encoded
between group and the project Id. I think the group name itself should be
encoded if it contains a slash within.
Unfortunately, I don't have time to work on it. Will be using my fork.
Thank you and good luck

On Wed, Mar 13, 2019, 9:46 AM Dale Hui <notifications@github.com> wrote:

> Hi, I don't fully understand you. I'm not inventing my own escaping here.
>
> The current approach of using regexs is not very clean. My original though
> was that we'd be adding backslash escaping on top of url encoding which is
> also dirty...
> I believe both of these dirty approaches can be avoided by using the
> escaped path.
>
> Please check out the following thread:
> https://stackoverflow.com/questions/39270633/why-does-url-parse-not-populate-url-rawpath#
>
> Ahh, yeah, it looks like we should be using EscapedPath() instead.
> https://play.golang.org/p/V838Xn_OZMa
>
> so pe := strings.Split(strings.Trim(u.Path, "/"), "/") would become pe :=
> strings.Split(strings.Trim(u.EscapedPath(), "/"), "/")
>
> current gitlab implementation is pretty much useless for 95% of us.
>
> No idea about the data for this... e.g. number of users using
> groups/projects Support for gitlab sources was only merged ~2 days ago.
> Obviously, the original author didn't have this problem.
>
> it would be super cool if you would approve it as it already is to unblock
> us all.
>
> If you're blocked, use your fork.
>
> Also, *whatever the final solution is, we should document it* so future
> users aren't confused.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/golang-migrate/migrate/pull/188#issuecomment-472314733>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Are_L0cJoUsjJBCmI0fNM16wiDl4c27-ks5vWKzvgaJpZM4bphdL>
> .
>
